services/wallet/database.go

```diff
@@ -0,0 +1,254 @@
+	db *sql.DB
+}
+
+// Close closes database.
+func (db Database) Close() error {
+	return db.db.Close()
+}
+
+// ProcessTranfers atomically adds/removes blocks and adds new tranfers.
+func (db Database) ProcessTranfers(transfers []Transfer, added, removed []*types.Header) (err error) {
+	var (
+		tx     *sql.Tx
+		insert *sql.Stmt
+		blocks *sql.Stmt
+		delete *sql.Stmt
+	)
+	tx, err = db.db.Begin()
+	if err != nil {
+		return err
+	}
+	defer func() {
+		if err == nil {
+			err = tx.Commit()
+			return
+		}
+		_ = tx.Rollback()
+	}()
+
+	insert, err = tx.Prepare("INSERT INTO transfers(hash, blk_hash, tx, receipt, type) VALUES (?, ?, ?, ?, ?)")
+	if err != nil {
+		return err
+	}
+	delete, err = tx.Prepare("DELETE FROM blocks WHERE hash = ?")
+	if err != nil {
+		return err
+	}
+	blocks, err = tx.Prepare("INSERT INTO blocks(hash,number,header) VALUES (?, ?, ?)")
+	if err != nil {
+		return err
+	}
+	for _, header := range removed {
+		_, err = delete.Exec(header.Hash())
+		if err != nil {
+			return err
+		}
+	}
+	for _, header := range added {
+		_, err = blocks.Exec(header.Hash(), (*SQLBigInt)(header.Number), &JSONBlob{header})
+		if err != nil {
+			return err
+		}
+	}
+
+	for _, t := range transfers {
+		_, err = insert.Exec(t.Transaction.Hash(), t.Header.Hash(), &JSONBlob{t.Transaction}, &JSONBlob{t.Receipt}, t.Type)
+		if err != nil {
+			return err
+		}
+	}
+	return err
+}
+
+// GetTransfers load transfers transfer betweeen two blocks.
+func (db *Database) GetTransfers(start, end *big.Int) (rst []Transfer, err error) {
+	query := "SELECT type, blocks.header, tx, receipt FROM transfers JOIN blocks ON blk_hash = blocks.hash WHERE blocks.number >= ?"
+	var (
+		rows *sql.Rows
+	)
+	if end != nil {
+		query += " AND blocks.number <= ?"
+		rows, err = db.db.Query(query, (*SQLBigInt)(start), (*SQLBigInt)(end))
+	} else {
+		rows, err = db.db.Query(query, (*SQLBigInt)(start))
+	}
+	if err != nil {
+		return
+	}
+	defer rows.Close()
+	for rows.Next() {
+		transfer := Transfer{
+			Header:      &types.Header{},
+			Transaction: &types.Transaction{},
+			Receipt:     &types.Receipt{},
+		}
+		err = rows.Scan(
+			&transfer.Type, &JSONBlob{transfer.Header},
+			&JSONBlob{transfer.Transaction}, &JSONBlob{transfer.Receipt})
+		if err != nil {
+			return nil, err
+		}
+		rst = append(rst, transfer)
+	}
+	return
+}
+
+// SaveHeader stores single header.
+func (db *Database) SaveHeader(header *types.Header) error {
+	_, err := db.db.Exec("INSERT INTO blocks(number, hash, header) VALUES (?, ?, ?)", (*SQLBigInt)(header.Number), header.Hash(), &JSONBlob{header})
+	return err
+}
+
+// SaveHeaders atomically stores list of headers.
+func (db *Database) SaveHeaders(headers []*types.Header) (err error) {
+	var (
+		tx     *sql.Tx
+		insert *sql.Stmt
+	)
+	tx, err = db.db.Begin()
+	if err != nil {
+		return
+	}
+	insert, err = tx.Prepare("INSERT INTO blocks(number,hash,header) VALUES (?,?,?)")
+	if err != nil {
+		return
+	}
+	defer func() {
```

Comment:
It is repeated a few times:
```suggestion
	defer func() { err = closeTx(tx) }() 
```

Reply:
i tried. it looks much more readable the way it is now. note that i have to ignore error if tx is rollbacked and i need to pass error to closeTx to detemine how to close it

Reply:
Ah right, it would look like `err = closeTx(tx, err)`. Not really clear.